### ISSUE TITLE
Add a draggable scale to maps

### ISSUE THREAD

**Issue by Futuray-pgm** [cgeo/cgeo#15425]
### Tell us your idea!

I think it will be realy useful to made the scale draggable around the map. 

### Additional Information

I know that it can be really hard to implement.

**Comment by MagpieFourtyTwo** [comment]
Maybe I'm missing something, but what exactly would be the useful part?

**Comment by brivars** [comment]
> Maybe I'm missing something, but what exactly would be the useful part?

I'm guessing, to estimate distances / do rough measurements. I have sometimes found myself trying to pan the map around in a way that the scale indicatir would roughly sit so I could guesstimate a distance between two caches.

However I'm not sure if a "draggable scale" would be the best solution. 
What would be better would be to add some sort of a ruler function, as found on Organic Maps, Google Maps, and other mapping apps.

**Comment by moving-bits** [comment]
To measure the distance between two caches, it's easier to put them into a so-called "individual route" (by long-tapping on them). c:geo will then display the distance between them (or, to be more specific: the length of the route), either as bee-line or routed for walking, bicycle or car, depending on your routing settings.

**Comment by Futuray-pgm** [comment]
I wanted that a scale like the one that already exist bottom of each map.

And it will be easier to measure between two points without to create two personalised caches.

**Comment by eddiemuc** [comment]
I imagine something like this:
- on map long-tap, add a new option to context Menü: "start measurement" (wording tbd)
- toast appear: "tap on target for measurement". Tapped start point will be marked b e.g. a cross symbol
- next tap on map will NOT trigger the normal action but instead draw a line from start tap to second tapped point with length written onto it (e.g. 530m or 3,2km)
- measurement line can be tapped to be deleted. Tapping will open a confirm dialog asking for line to be deleted 
- nothing of the above will be persisted in any way. Leaving map and reentering will remove measurements

**Comment by Futuray-pgm** [comment]
Yes!
And when clicking the measurement, a context menu appears with this options:
- "`change color`"
- "`informations`"
- "`save measurement`" (into the current list)
- "`save all measurements`"
- "`delete measurement`"
- "`delete all measurements`"

**Comment by fm-sys** [comment]
IMHO cgeo should focus on geocaching. While measurements might be a small nice additional feature, it should be nothing more than that. Focus on simplicity!

- on map long-tap, select start measurement (that paint is used as start)
- user can do the measurement by moving the map around (center is used as target point)
- pressing back button key removes it again (no confirmation necessary)

Changing colors, storing multiple lines etc goes far beyond that scope, and can already be archived by using GPX / individual route. c:geo is no tool for doing map paintings :-)

**Comment by eddiemuc** [comment]
> Changing colors, storing multiple lines etc goes far beyond that scope, and can already be archived by using GPX / individual route. c:geo is no tool for doing map paintings :-)

Side note: this can already be achieved today for polylines, polygons and circles using all colors of the rainbow by importing/displaying custom GeoJson-Files.

**Comment by ztNFny** [comment]
the existing individial route feature already allows adding 2 arbitrary points (caches or non-caches) and shows the distance between them. Doesn't that already 100% cover this requirement?

![image](https://github.com/cgeo/cgeo/assets/1258173/45d4ef9a-a161-40ee-b03a-00ace8df45ad)

**Comment by Futuray-pgm** [comment]
Yes, but we can only have one route. And we can't compare.

**Comment by ztNFny** [comment]
> Yes, but we can only have one route. And we can't compare.

the same applies to your proposed solution (draggable scale marker)?!

**Comment by eddiemuc** [comment]
> the existing individial route feature already allows adding 2 arbitrary points (caches or non-caches) and shows the distance between them. Doesn't that already 100% cover this requirement?

Downside is that you would have to delete any existing individual route just to use this feature to measure between two points.

I actually like @fm-sys s suggestion from above (better than my own)

**Comment by ztNFny** [comment]
> Downside is that you would have to delete any existing individual route just to use this feature to measure between two points.

How likely is it that you need a measurement that's completly unrelated to your route? (If it was related to the route then appending it to the route would be exactly what you want). And even then you can just append them to route. Add point 1, remember the number, add point 2, substract, done.
I see the value added by a seperate menu item way too small compared to the added UI clutter.

**Comment by eddiemuc** [comment]
I personally don't use the route features at all, but would appreciate the measurement feature. Routing GUI is too heavy-weighted for that IMHO (add first point, add last point, read measurement, delete route again, do the same procedure to have a measurement to a slightly different located point, ...). Compared to "start measurement to start, then drag around as I want, then stop measurement again)

**Comment by ztNFny** [comment]
But what's the value of knowing the distance between 2 points in a straight line? That's only very seldomly the relevant information - paths are not straight. So in real life you'll want either "distance from A via B to C" or "distance from A using the path to C" - both something that the suggested solutions don't provide but routing already does.
And the extra click required in routing is deleting the route at the end (from your mentioned steps "add first point", "add last point" (at least my moving the map), "read measurement" would also be required for the suggested solution) - for the second measurement no need to start anew but just remove point B and add another one..
Imho the suggested solution of "measure a straight line between 2 points" will not bring any value, yet clutter the UI with yet another (8th) menu item of which each user probably uses a maximum of 2 (different 2 per user).

**Comment by Futuray-pgm** [comment]
> "read measurement" would also be required for the suggested solution

No, it will be written in center of the line, or top-left of screen.

**Comment by eddiemuc** [comment]
> But what's the value of knowing the distance between 2 points in a straight line? 

This is good enough for me. I usually don't even have routing information downloaded because I simply don't need it. I would argue that routing takes up way too much space and GUI in c:geo as of today.

The value of @fm-sys suggestion above IMHO is its dynamic: you can use dragging to quickly get distances between one ooint and others. Including zooming etc. That flexibility is not reachable via routing functions w/o constantly adding and then deleting the second route point.

> ...yet clutter the UI with yet another (8th) menu item of which each user probably uses a maximum of 2 (different 2 per user).

While this is certainly true, we have to stop bringing this argument up for new feature request. Rather we should take steps to clean up the GUI as a while and reduce functionality drastically. Taking the mentioned menu as an example:

![image](https://github.com/cgeo/cgeo/assets/6909759/e74d3c4c-5cf4-411d-af0c-b7c7e3cb8614)

* when I long-tap on a clear spot on the map I have not less than 4 (!) menu entries related to navigation: append to route, prepend to route, set as target and navigate. That's way too cluttered, and also hard to understand what the different items will do if tapped (to be honest: I too don't know the meaning of "set as target"). Why not make a submenu out of it?
* When routing information is available I have a dummy entry "elevation (map)" in the menü. Why is this placed in a submenu? Could be merged with the "show coordinate" information.

**Comment by ztNFny** [comment]
> While this is certainly true, we have to stop bringing this argument up for new feature request. Rather we should take steps to clean up the GUI as a while and reduce functionality drastically. Taking the mentioned menu as an example:

Adding functionality vs reducing functionality?
Every feature, old and new, needs to be evaluated for the value of brings. So of course the argument needs to be made for every new FR!
The problem is that removing something is a lot harder than not adding it - because no matter what you remove, there's at least 1 person out there for whom this feature is important.

> * when I long-tap on a clear spot on the map I have not less than 4 (!) menu entries related to navigation: append to route, prepend to route, set as target and navigate. That's way too cluttered, and also hard to understand what the different items will do if tapped (to be honest: I too don't know the meaning of "set as target"). Why not make a submenu out of it?

Navigate it's not related to routing fwiw.

> * When routing information is available I have a dummy entry "elevation (map)" in the menü. Why is this placed in a submenu? Could be merged with the "show coordinate" information.

See https://github.com/cgeo/cgeo/pull/15429 😉

**Comment by ztNFny** [comment]
If the team agrees that this feature brings enough value then it should be added. But there needs to be room to have that discussion. A discussion with pros, cons and different perspectives (all of which are opinions, not truths!) that then need to be weighed and a decision be made upon.

Otherwise we'll end up with another "why are there 4 menu items related to routing, they should ask be removed" (sorry for the simplification, just picked up the obvious example from above) soon.

**Comment by ztNFny** [comment]
> to be honest: I too don't know the meaning of \"set as target\

![Screenshot_2024-03-09-16-56-58-48_a916ceb0c8a82619faeb76fe6b8ae997.jpg](https://github.com/cgeo/cgeo/assets/1258173/3f1be8ea-18ea-4225-95f6-6d509927e370)

😉

**Comment by eddiemuc** [comment]
> there needs to be room to have that discussion. A discussion with pros, cons and different perspectives (all of which are opinions, not truths!) that then need to be weighed and a decision be made upon.

Fully agree. I think this is what we're doing right now and I think it works.

**Comment by eddiemuc** [comment]
> > to be honest: I too don't know the meaning of \"set as target\
> 
> ![Screenshot_2024-03-09-16-56-58-48_a916ceb0c8a82619faeb76fe6b8ae997.jpg](https://github.com/cgeo/cgeo/assets/1258173/3f1be8ea-18ea-4225-95f6-6d509927e370)
> 
> 😉

I still don't fully get it... I played around a bit with "set as target" vs "append/prepend route". It seems as if I can kind of create two different "routes": one with just one "target" and one route with many waypoints. I even get two different distance/length displays in the map for the two. How do those two functions differ?

**Comment by moving-bits** [comment]
"Set as target" allows to easily set an arbitrary point on the map as current target ("pink direction line"), without having to create a new or touch an existing route. (This is sort of a shortcut to select coordinate - navigate to - map)

"Append/prepend to route" adds this coordinate to a route ("blue line").

**Comment by ztNFny** [comment]
> > > to be honest: I too don't know the meaning of \"set as target\
> > 
> > ![Screenshot_2024-03-09-16-56-58-48_a916ceb0c8a82619faeb76fe6b8ae997.jpg](https://github.com/cgeo/cgeo/assets/1258173/3f1be8ea-18ea-4225-95f6-6d509927e370)
> > 
> > 😉
> 
> I still don't fully get it... I played around a bit with "set as target" vs "append/prepend route". It seems as if I can kind of create two different "routes": one with just one "target" and one route with many waypoints. I even get two different distance/length displays in the map for the two. How do those two functions differ?

I think this is getting a bit off topic here, but:
There's one target, red line. This is what pretty much every user should be familiar with, leading you to the next geocache.
Then there's Route with multiple waypoints (blue line). This can be used for planning ahead. 
Example use case: you add 10 caches to route, you will see that the total is 5 km. The first geocache is set as target, you get a red line to it and in four it's in 100 meters. After finding that geocach you remove it from the route and the next cache in the route will be selected as target.

(If that's not enough of an explanation I'd suggest to move this to tomorrows call. I'm also sure there's some written documentation somewhere, but I'm not so familiar with the wiki)

**Comment by eddiemuc** [comment]
> I think this is getting a bit off topic here, but: There's one target, red line. This is what pretty much every user should be familiar with, leading you to the next geocache. Then there's Route with multiple waypoints (blue line). This can be used for planning ahead. Example use case: you add 10 caches to route, you will see that the total is 5 km. The first geocache is set as target, you get a red line to it and in four it's in 100 meters. After finding that geocach you remove it from the route and the next cache in the route will be selected as target.

Thanks for explanation, I believe I got it now...

Trying to get back to the topic... (:-)): isn't "set as target" then much like "prepend to route" in that it:
* if no route exists: creates a "route" from current position to target
* if a route already exists: "prepends" the waypoint to the route

If this is true can't one of the two entries be removed? ("Set as target" or "prepend to route")
--> and thus make room for a "start measurement" entry? :-)

**Comment by moving-bits** [comment]
> Trying to get back to the topic... (:-)): isn't "set as target" then much like "prepend to route" in that it:

no, as "set to target" does not touch a "individual route" at all (neither add to an existing one nor create a new one" (see https://github.com/cgeo/cgeo/issues/15425#issuecomment-1986900632)

**Comment by eddiemuc** [comment]
> > Trying to get back to the topic... (:-)): isn't "set as target" then much like "prepend to route" in that it:
> 
> no, as "set as target" does not touch a "individual route" at all (neither add to an existing one nor create a new one" (see [#15425 (comment)](https://github.com/cgeo/cgeo/issues/15425#issuecomment-1986900632))

le me rephrase: **shouldn't** "set as target" be modified that it acts as if "prepend to route" is used? I don't understand why one would need both concepts in parallel

**Comment by moving-bits** [comment]
ah, now I understand :-)

I guess it depends: We still have users that have never heard of "individual routes" (which leads to questions like "what is this blue line" or "what are those numbers on the map for"). If you don't have an individual route yet, and "set as target" would "add (or prepend) to (empty) route", seemingly nothing would happen, as an individual route would need two items to show something on the map (blue line and distance). Except if you would also switch "set route start as target" to "on" automatically in this case.

Right now it's just an easy way to route to a user-selected target without touching the route at all.

**Comment by MagpieFourtyTwo** [comment]
>We still have users that have never heard of "individual routes"

Even worse - sometimes it's kind of mind boggling how inexperienced inexperienced users are. Just last night one of this kind asked me at an event why her map's context menu suddenly stopped to work (BTW: has the default for ``Long tap on map`` changed?). And while explaining the setting to her it turned out that she did not even know that c:geo is capable of routing (resp. asking BRouter for a route ;) ). And guess what: I just started to explain and configure the internal routing, three others were standing around us, wanting to know how this "cool new feature" works ... eventually I ended up with five routing installations and configurations ... five out of ten guys at our table!  That makes 50 %. Actually even more, because at least three of the other 50 % were Apple enthusiasts. ;)

However, back to topic: I really love the wide range of features c:geo offers, and it's great that everyone can cover their requirements and put their own "toolbox" together, but we should nevertheless be aware to not overload it. Apart from the increasing complexity, which at some point becomes almost unmanageable and makes the whole thing increasingly difficult to maintain, it also becomes more and more difficult for the users to find what they really need resp. what they are looking for.

TMALSS: I would not recommend to implement too many features, and esp. not a second way for a functionality which can already be achieved with the already given stuff. Like with "draggable scale". Most of the users don't even know how to make a "standard" route, and even less how to use individual routes. There still are even guys who do not even know about the map's long tap menu. How low will the percentage be for a draggable scale ...? Esp. as it already can be achieved with two arbitrary points. And believe me, even these are already hard enough to explain ...  ;)

So I go with @fm-sys : Focus on simplicity. KISS - keep it simple and stupid.

**Comment by MagpieFourtyTwo** [comment]
By the way, this "inexperienced" user is not even an occasional one, she is using c:geo almost day and night 24/7/365 and is caching an average of around 6000 stashes per year - and has been doing so for several years now! All with c:geo. And yet she doesn't really know how to use it to its full potential ... ;)  So guess, how all this may be for those who use it on weekends only.

**Comment by mapa6** [comment]
Hi, it would be nice add measure distance function for two fingers long touch on map (same function is in mapy.com app include distance and azimuth). Its simply to use. Try it...

**Comment by MagpieFourtyTwo** [comment]
> Hi, it would be nice add measure distance function for two fingers long touch on map (same function is in mapy.com app include distance and azimuth). Its simply to use. Try it...

As mentioned in the thread before, this is already possible.  Only difference is that after you have done the first touch, you have to lift your finger and then touch the second point.  Even better: If the starting point is your current position, you don't even need the first touch.  Plus of course you should have set the route to "straight" for bee line (which I would say is pretty useless, cause only reason for getting the distance is to estimate how much effort it will be to reach the destination) or, what should always be the case anyway, to you current mode of transport - with which you even get way more sense-full information like the real walking distance.  Plus even an elevation chart.
